streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -904,6 +869,83 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
 }
 
 object ZSink extends ZSinkPlatformSpecific {
+
+  implicit class InvariantOps[R, E, A, B](val sink: ZSink[R, E, A, A, B]) extends AnyVal {
+
+    /**
+     * Returns a new sink that tries to produce the `B`, but if there is an
+     * error in stepping or extraction, produces `None`.
+     */
+    final def ? : ZSink[R, Nothing, A, A, Option[B]] =
+      new ZSink[R, Nothing, A, A, Option[B]] {
+        type State = Option[sink.State]
+
+        val initial = sink.initial.map(Step.leftMap(_)(Option(_))) orElse
+          UIO.succeed(Step.done(None, Chunk.empty))
+
+        def step(state: State, a: A) =
+          state match {
+            case None => UIO.succeed(Step.done(state, Chunk.empty))
+            case Some(state) =>
+              sink
+                .step(state, a)
+                .foldM(
+                  _ => UIO.succeed(Step.done(None, Chunk.single(a))),
+                  s => UIO.succeed(Step.leftMap(s)(Some(_)))
+                )
+          }
+
+        def extract(state: State) =
+          state match {
+            case None        => UIO.succeed(None)
+            case Some(state) => sink.extract(state).map(Some(_)) orElse UIO.succeed(None)
+          }
+      }
+
+    /**
+     * A named alias for `?`.
+     */
+    final def optional: ZSink[R, Nothing, A, A, Option[B]] = ?
+  }
+
+  implicit class NoRemainderOps[R, E, A, B](val sink: ZSink[R, E, Nothing, A, B]) extends AnyVal {
+
+    /**
+     * Returns a new sink that tries to produce the `B`, but if there is an
+     * error in stepping or extraction, produces `None`.
+     */
+    final def ? : ZSink[R, Nothing, A, A, Option[B]] =
+      new ZSink[R, Nothing, A, A, Option[B]] {
+        type State = Option[sink.State]
+
+        val initial = sink.initial.map(Step.leftMap(_)(Option(_))) orElse
+          UIO.succeed(Step.done(None, Chunk.empty))
+
+        def step(state: State, a: A) =
+          state match {
+            case None => UIO.succeed(Step.done(state, Chunk.empty))
+            case Some(state) =>
+              sink
+                .step(state, a)
+                .foldM(
```

Comment:
It looks like you could just use `fold` and drop the `UIO.succeed` here.

Reply:
Thanks. Done.